### ISSUE TITLE
Unexpected EOF when receiving emails from Steam

### ISSUE THREAD

**Issue by axtgr** [foxcpp/maddy#300]
# Describe the bug

I'm unable to receive any emails from Steam because of the "unexpected EOF" error. 

# Steps to reproduce

Make Steam send you an email (e.g. buy a game or sell an item on the market).

# Log files

```
mail_1  | [debug] smtp/pipeline: initializing state for require_mx_record: (0xc0001a4640)                                                                                                                                             {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: sender noreply@steampowered.com matched by default rule                                                                                                                                              {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: global rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                            {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: per-source rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                        {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: recipient mail@example.com matched by domain rule 'example.com'                                                                                                                                      {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: per-rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                               {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: sender noreply@steampowered.com matched by default rule                                                                                                                                                {"msg_id":"4aba6493"}
mail_1  | [debug] smtp/pipeline: tgt.Start(noreply@steampowered.com) ok, target = msgpipeline:local_routing                                                                                                                           {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: global rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                              {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: per-source rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                          {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: recipient mail@example.com matched by domain rule 'example.com'                                                                                                                                        {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: per-rcpt modifiers: mail@example.com => mail@example.com                                                                                                                                                 {"msg_id":"4aba6493"}
mail_1  | [debug] msgpipeline: tgt.Start(noreply@steampowered.com) ok, target = sql:local_mailboxes                                                                                                                                   {"msg_id":"4aba6493"}
mail_1  | smtp: RCPT ok        {"msg_id":"4aba6493","rcpt":"mail@example.com"}
mail_1  | [debug] autobuffer: keeping the message in RAM
mail_1  | smtp: DATA error     {"msg_id":"4aba6493","reason":"unexpected EOF"}
mail_1  | [debug] smtp: reset
```

# Configuration file

Basically the default config with multiple domains set up + DKIM and SPF checks removed from smtp as they were apparently causing other issues.

# Environment information

* maddy version 0.4.2 
* Docker image foxcpp/maddy@latest

**Comment by foxcpp** [comment]
Seems like an issue at SMTP protocol level. Could you enable I/O tracing for port 25 endpoint and give me logs? (they will likely include the message itself or part of it)
To enable it, add `io_debug` and `debug` to `smtp` block in configuration. 
```
smtp tcp://0.0.0.0:25 {
    debug
    io_debug
    ...
}
```

**Comment by ghost** [comment]
Same issue while receiving GitHub’s verify email. It can be reproduced by sending or receiving a big email (>3MB). 
I can temporarily solve this problem by adding `buffer ram` into `smtp` block and `submission` block.

Are there something wrong in `autobuffer`?

Environment: Docker image foxcpp/maddy@latest

**Comment by axtgr** [comment]
@foxcpp here is the log: https://hastebin.com/filonogiho

**Comment by foxcpp** [comment]
925b758 should fix the bug.

**Comment by axtgr** [comment]
That's great, thanks!

In the initial post I also mentioned that I was having issues with the DKIM and SPF checks with emails from Steam. They were probably caused by the same bug and are fixed now, but here is the log just in case: https://hastebin.com/welohubeji

**Comment by foxcpp** [comment]
Closing as it was fixed. Fix is available on master branch and will be part of 0.4.3 release.

**Comment by daliusd** [comment]
I have experienced same problem with gravatar (wordpress) login e-mail with version 0.4.2. It is working fine with master version.
